src/Adapter/Shipment/OrderShipmentCreator.php

```diff
@@ -77,17 +77,43 @@ public function addShipmentOrder(Order $order, array $productsHandledByCarrier):
             // match products with order details to get quantities & orderDetailId
             foreach (OrderDetail::getList($order->id) as $orderDetailProduct) {
                 foreach ($products['product_list'] as $product) {
-                    if ($product['id_product'] === $orderDetailProduct['product_id']) {
-                        $shipmentProduct = new ShipmentProduct();
-                        $shipmentProduct->setShipment($shipment);
-                        $shipmentProduct->setOrderDetailId($orderDetailProduct['id_order_detail']);
-                        $shipmentProduct->setQuantity($orderDetailProduct['product_quantity']);
-                        $shipment->addShipmentProduct($shipmentProduct);
+                    if (!$this->needShipmentProductCreation($product, $orderDetailProduct)) {
+                        continue;
                     }
+
+                    $shipmentProduct = (new ShipmentProduct())
+                        ->setShipment($shipment)
+                        ->setOrderDetailId($orderDetailProduct['id_order_detail'])
+                        ->setQuantity($orderDetailProduct['product_quantity']);
+
+                    $shipment->addShipmentProduct($shipmentProduct);
                 }
             }
 
             $this->shipmentRepository->save($shipment);
         }
     }
+
+    /**
+     * @param array{
+     *     id_customization: int,
+     *     id_product: int
+     * } $product
+     * @param array{
+     *     id_customization: int,
+     *     id_order_detail: int,
+     *     product_id: int,
+     *     product_quantity: int
+     * } $orderDetailProduct
+     *
+     * @return bool
+     */
+    private function needShipmentProductCreation(array $product, array $orderDetailProduct): bool
+    {
+        if (!empty($product['id_customization'])) {
```

Comment:
I think that you should be matching by both id_product, id_product_attribute and id_customization.

But it's still a code smell and a potential point of failure. It should be refactored before going into production imho.